subsys/net/lib/lwm2m/lwm2m_registry.c

```diff
@@ -1344,3 +1355,324 @@ bool lwm2m_engine_shall_report_obj_version(const struct lwm2m_engine_obj *obj)
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (path.level < 3) {
+		LOG_ERR("path must have at least 3 parts");
+		return -EINVAL;
+	}
+
+	/* look up resource obj */
+	ret = path_to_objs(&path, &obj_inst, &obj_field, NULL, &res_inst);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (!res_inst) {
+		LOG_ERR("res instance %d not found", path.res_inst_id);
+		return -ENOENT;
+	}
+
+	switch (obj_field->data_type) {
+	case LWM2M_RES_TYPE_U32:
+	case LWM2M_RES_TYPE_TIME:
+	case LWM2M_RES_TYPE_U16:
+	case LWM2M_RES_TYPE_U8:
+	case LWM2M_RES_TYPE_S64:
+	case LWM2M_RES_TYPE_S32:
+	case LWM2M_RES_TYPE_S16:
+	case LWM2M_RES_TYPE_S8:
+	case LWM2M_RES_TYPE_BOOL:
+	case LWM2M_RES_TYPE_FLOAT:
+		/* Support only fixed width resource types */
+		cache_entry = lwm2m_cache_entry_allocate(resource_path);
+		break;
+	default:
+		cache_entry = NULL;
+		break;
+	}
+
+	if (!cache_entry) {
+		return -ENODATA;
+	}
+
+	ring_buf_init(&cache_entry->rb, cache_entry_size * cache_len, (uint8_t *)data_cache);
+
+	return 0;
+#else
+	LOG_ERR("LwM2M resource cache is only supported for "
+		"CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT");
+	return -ENOTSUP;
+#endif /* CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT */
+}
+
+int lwm2m_engine_data_cache_init(void)
+{
+#if defined(CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT)
+	int i;
+
+	sys_slist_init(&lwm2m_timed_cache_list);
+
+	for (i = 0; i < ARRAY_SIZE(lwm2m_cache_entries); i++) {
+		lwm2m_cache_entries[i].path = NULL;
+	}
+#endif
+	return 0;
+}
+
+struct lwm2m_time_series_elem *lwm2m_cache_alloc(struct lwm2m_time_series_resource *cache_entry)
+{
+#if defined(CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT)
+	uint32_t length;
+	uint8_t *buf_ptr;
+	uint32_t element_size = sizeof(struct lwm2m_time_series_elem);
+
+	if (ring_buf_space_get(&cache_entry->rb) < element_size) {
+		/* No space  */
+		if (IS_ENABLED(CONFIG_LWM2M_CACHE_DROP_LATEST_DATA)) {
+			return NULL;
+		}
+
+		length = ring_buf_get_claim(&cache_entry->rb, &buf_ptr, element_size);
+		ring_buf_get_finish(&cache_entry->rb, length);
+		LOG_ERR("Drop oldest data");
+	}
+
+	length = ring_buf_put_claim(&cache_entry->rb, &buf_ptr, element_size);
+
+	if (length != element_size) {
+		ring_buf_put_finish(&cache_entry->rb, 0);
+		LOG_ERR("Allocation failed %u", length);
+		return NULL;
+	}
+
+	ring_buf_put_finish(&cache_entry->rb, length);
+	return (struct lwm2m_time_series_elem *)buf_ptr;
+#else
+	return NULL;
+#endif
+}
+
+
+struct lwm2m_time_series_elem *lwm2m_cache_read(struct lwm2m_time_series_resource *cache_entry)
+{
+#if defined(CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT)
+	uint32_t length;
+	uint8_t *buf_ptr;
+	uint32_t element_size = sizeof(struct lwm2m_time_series_elem);
+
+	if (ring_buf_is_empty(&cache_entry->rb)) {
+		return NULL;
+	}
+
+	length = ring_buf_get_claim(&cache_entry->rb, &buf_ptr, element_size);
+
+	if (length == element_size) {
+		ring_buf_get_finish(&cache_entry->rb, length);
+		return (struct lwm2m_time_series_elem *)buf_ptr;
```

Comment:
Is there a problem here?
You call `ring_buf_get_finish()` but then access the memory you claimed initially?

If there is a sensor simultaneously writing into a buffer, we might allocate the same space.

Reply:
When I add registry mutex to lwm2m_engine_send() all ring buffer read and write operations are saved by mutex.